### PR TITLE
Disconnect cleanups, idle timeouts, auto heartbeats

### PR DESCRIPTION
**Pull request by RJ**
* Added `NetworkResource.disconnect(handle)` which cleans up internal structures.
* Added `NetworkingPlugin.idle_timeout_ms` which, if set, will drop connections if no packets received for this long
* Added `NetworkingPlugin.auto_heartbeat_ms` which, if set, will send a heartbeat packet as needed. Heartbeat packets are not propagated to user-provided bevy systems. They just exist to be received and dropped, to update the idle timeout and keep us from being dropped.
* Added `idle_timeout` example and readme.
* Revamped argparser with a dev-dependency on clap.

The heartbeats and idle timeouts are checked in a bevy system on a fixed timestep much less frequently than per-tick, since that seemed wasteful.

#### When a client times out
If you stop a client, wait for the server to drop and cleanup the connection after `idle_timeout_ms`, then restart the client, it will connect with a new `ConnectionHandle`.

#### When you restart a client quickly without a timeout
If you restart a client quickly, it will reconnect with the same `ConnectionHandle`. This behaviour is unchanged, but it feels like a bug. I think it happens because we identify clients by `SocketAddr`, ie a combo of IP+port. Typically as a client, you'd let the operating system's ephemeral port system assign you a local port, sometimes by specifying a port of 0 (eg: `src_addr = "127.0.0.1:0"`). However, `naia` uses a [linear search for available ports](https://github.com/naia-rs/naia-socket/blob/master/shared/src/find_available_port.rs) so I presume it is reselecting the same port when you quickly restart a client. IMO this would be better left for the OS to pick, unless there is a good reason for this method i'm unaware of.

#### Higher-level solution
Another way to handle this is at the protocol level, where we exchange a random session token as part of a handshake. Probably out of scope for this library though?

Related issues: #6 and #17

### CODE REVIEW

File: src/transport.rs [hunk 1]
```diff
@@ -33,6 +33,8 @@ pub type MultiplexedPacket = MuxPacket<<BufferPacketPool<SimpleBufferPool> as Pa
 pub type ConnectionChannelsBuilder =
     MessageChannelsBuilder<TaskPoolRuntime, MuxPacketPool<BufferPacketPool<SimpleBufferPool>>>;
 
+pub const HEARTBEAT_PACKET: &[u8] = b"BNT:HEARTBEAT";
```
Comment (smokku) [inline src/transport.rs:36]: Heartbeat is usually done using an empty packet. There is no need it has any actual content.

**Comment by smokku** [comment]
One of the reasons I didn't tackle disconnection problem is that it needs a protocol. There is a need to:

1. tell the client: you are good and connected
2. you are not wanted anymore, stop sending data
3. you sent a packet I didn't want and it was dropped

Basically what SYN/ACK/RST of TCP does.

I did not come up wit a solution that does not interfere with an option that you are free to send RAW packets of your own protocol.

File: src/transport.rs (hunk 1 above)
Comment (RJ) [reply src/transport.rs]: ok

File: src/lib.rs [hunk 2]
```diff
@@ -389,6 +446,38 @@ impl NetworkResource {
...
+            silent_handles.push(*handle);
+        }
+        if heartbeat_limit.is_some() && tx_ms > heartbeat_limit.unwrap() as u128 {
+            // send a heartbeat packet.
+            needs_hb_handles.push(*handle);
+        }
+    }
+    for handle in needs_hb_handles {
+        log::debug!("Sending hearbeat packet on h:{}", handle);
+        net.send(handle, Packet::from_static(transport::HEARTBEAT_PACKET)).unwrap();
```
Comment (smokku) [inline src/lib.rs:470]: `Packet::new()` maybe?

File: src/lib.rs [hunk 3]
```diff
@@ -415,6 +504,11 @@ pub fn receive_packets(
         while let Some(result) = connection.receive() {
             match result {
                 Ok(packet) => {
+                    if packet == transport::HEARTBEAT_PACKET {
```
Comment (smokku) [inline src/lib.rs:507]: `if packet.len() == 0` maybe?

**Comment by RJ** [comment]
Yeah wasn't sure if handshakey bits belong in this library or not really.

I'm leaning towards yes, because as a consumer of this crate i just want the connection setup and teardown to be abstracted away for me i think. 

We could open a reliable message channel for signaling things like a handshake, disconnect, etc. afaik that wouldn't conflict with being able to send raw packets. We could defer sending `NetworkEvent::Connected` until our handshake stuff is complete. If you think this could work, i'd be happy to try it out once we agree what it should look like.

File: src/lib.rs (hunk 2 above)
Comment (RJ) [reply src/lib.rs]: i removed HEARTBEAT_PACKET entirely

**Comment by RJ** [comment]
I think this PR as-is is a useful incremental improvement, even without a protocol for handshaking/disconnecting. It means you can restart clients and connect to the same server-instance without a problem, provided the server has timed them out.  Makes client/server dev/test locally a bit faster.

**Comment by smokku** [comment]
> I'm leaning towards yes, because as a consumer of this crate i just want the connection setup and teardown to be abstracted away for me i think.

Agreed.

> We could open a reliable message channel for signaling things like a handshake, disconnect, etc. afaik that wouldn't conflict with being able to send raw packets. We could defer sending `NetworkEvent::Connected` until our handshake stuff is complete. If you think this could work, i'd be happy to try it out once we agree what it should look like.

It can be hidden under the message channels setup. Together with user provided channels you set up a signaling channels.
This way until user explicitly sets up channels configuration one is free to send whatever RAW packet one wants. After setting channels up, turbulence takes over and channels (together with the signaling one) are the only way to go.

File: src/lib.rs [hunk 4]
```diff
@@ -229,8 +268,8 @@ impl NetworkResource {
                                 channel.send(Ok(Packet::copy_from_slice(packet.payload())))
                             }) {
                             Some(Ok(())) => false,
-                            Some(Err(error)) => {
-                                log::error!("Server Send Error: {}", error);
+                            Some(Err(CrossbeamSendError(_packet))) => {
```
Comment (smokku) [inline src/lib.rs:271]: If it is used it should be `packet` not `_packet`.

File: src/transport.rs [hunk 5]
```diff
@@ -33,6 +33,7 @@ pub type MultiplexedPacket = MuxPacket<<BufferPacketPool<SimpleBufferPool> as Pa
 pub type ConnectionChannelsBuilder =
     MessageChannelsBuilder<TaskPoolRuntime, MuxPacketPool<BufferPacketPool<SimpleBufferPool>>>;
 
+
```
Comment (smokku) [inline src/transport.rs:36]: Unneeded.

File: src/transport.rs [hunk 6]
```diff
@@ -115,6 +116,7 @@ pub struct ServerConnection {
 
     channels: Option<MessageChannels>,
     channels_rx: Option<IncomingMultiplexedPackets<MultiplexedPacket>>,
+    #[allow(dead_code)]
```
Comment (smokku) [inline src/transport.rs:119]: Why is this needed?

**Review by smokku — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 4 above)
Comment (RJ) [reply src/lib.rs:271]: it's not used in the log message any more. it's a rare case as long as dropped connections are cleaned up, and the packet isn't lost - it's delivered on a newly created channel. so it seemed spammy to log the packet contents.

File: examples/utils/mod.rs [hunk 7]
```diff
@@ -1,40 +1,174 @@
...
+        .args(server_or_client_args().as_slice())
+        .get_matches();
+    SimpleArgs {
+        is_server: matches.is_present("server")
+    }
+}
+
+#[allow(dead_code)]
+pub fn parse_message_coalescing_args() -> MessageCoalescingArgs {
+    let matches = ClapApp::new("")
```
Comment (smokku) [inline examples/utils/mod.rs:53]:
```suggestion
    let matches = ClapApp::new(exe_name())
```

File: examples/utils/mod.rs [hunk 8]
```diff
@@ -1,40 +1,174 @@
...
+        pings: value_t_or_exit!(matches, "pings", usize),
+        pongs: value_t_or_exit!(matches, "pongs", usize),
+        manual_flush: matches.is_present("manual-flush"),
+        auto_flush: matches.is_present("auto-flush"),
+    }
+}
+
+#[allow(dead_code)]
+pub fn parse_idle_timeout_args() -> IdleTimeoutArgs {
+    let matches = ClapApp::new("")
```
Comment (smokku) [inline examples/utils/mod.rs:70]:
```suggestion
    let matches = ClapApp::new(exe_name())
```

File: src/transport.rs (hunk 6 above)
Comment (RJ) [reply src/transport.rs:118]: to silence this warning:

```
warning: field is never read: `channels_task`
   --> src/transport.rs:121:5
    |
121 |     channels_task: Option<Task<()>>,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

there is a spawned task assigned to it in build_channels, but we never read from it. looks like that's just to keep the task in scope so it's not dropped. is there a better way to make the compiler happy?

File: src/lib.rs (hunk 4 above)
Comment (smokku) [reply src/lib.rs:271]: Oh. Right. I misunderstood the change.

**Review by smokku — CHANGES_REQUESTED** [review]

File: src/transport.rs (hunk 6 above)
Comment (smokku) [reply src/transport.rs:118]: Oh yeah. Right.

I wish it was a way to explicitly mark this type of use, as `dead_code` suggests that this can be removed.
We need to add a comment explaining that this field is used for reference keeping and cannot be removed.

Comment (RJ) [reply src/transport.rs:118]: will do

File: src/transport.rs [hunk 9]
```diff
@@ -230,6 +234,7 @@ pub struct ClientConnection {
...
```
Comment (smokku) [inline src/transport.rs:237]: Here we need comment too.

**Review by smokku — CHANGES_REQUESTED** [review]

File: src/transport.rs (hunk 9 above)
Comment (RJ) [reply src/transport.rs:240]: ah yes, thanks. amended to add comment in both places

**Review by smokku — APPROVED** [review]

**Comment by smokku** [comment]
You should be able to merge this change once CI completes.
Please do a squash-merge as 8 commits is a bit much for one feature. ;-)

**Comment by RJ** [comment]
great, thanks :)
